### DISCUSSION TITLE
[21830] Minimum fragment size in SHM of Fast-DDS

### DISCUSSION THREAD

**Discussion by AnoopEmbeddedEngineer** [eProsima/Fast-DDS#5291] [General]
When i tried setting the maxmessagesize as 123 bytes in SHM through xml file,it shows RTPSWriter error pointing towards add_data() in the log..

Why that log message is coming...

What is the minimum fragment size or maxmessagesize that can be set in SHM ?

**Comment by AnoopEmbeddedEngineer** [comment]
![Screenshot from 2024-10-03 11-59-55](https://github.com/user-attachments/assets/5f783ba7-7b6a-47c6-9b76-b31909e9a532)

![Screenshot from 2024-10-03 12-00-40](https://github.com/user-attachments/assets/8fe278e3-7c0f-4bef-a0b2-26e49d062682)



When i set the maxmessage size to 123 bytes and started sending 1GB data, fast dds is showing this error..

**Reply by paxifaer** [reply]
Did you do this in the container?

**Reply by AnoopEmbeddedEngineer** [reply]
> Did you do this in the container?

No

**Reply by paxifaer** [reply]
That may be related to your user permissions. Please first check whether your program has sufficient permissions to request allocation. Generally, the shm size of dds depends on your machine.

**Reply by paxifaer** [reply]
123 bytes and started sending 1GB data. I'm a little unclear about your approach. You sliced ​​1GB of data yourself and each slice is <=123 bytes. Did you call data_write() at the end to write the data?

**Reply by AnoopEmbeddedEngineer** [reply]
> 123 bytes and started sending 1GB data. I'm a little unclear about your approach. You sliced ​​1GB of data yourself and each slice is <=123 bytes. Did you call data_write() at the end to write the data?

I was experimenting...
Even if i tried to send 64 MB under the same criteria, the issue occurs...

Do you know the minimum MaxMessageSize for SHM transport in fast dds?

**Reply by paxifaer** [reply]
My impression is that shm has no maximum limit.

**Reply by AnoopEmbeddedEngineer** [reply]
> My impression is that shm has no maximum limit.

Hi, I am asking what is the minimum limit for MaxMessageSize  in fast dds

**Reply by JesusPoderoso** [reply]
The recommended minimum limit for MaxMessageSize is 512B @Employeeprojec.

**Reply by AnoopEmbeddedEngineer** [reply]
> The recommended minimum limit for MaxMessageSize is 512B @Employeeprojec.

But when i tried to set MaxMessageSize as 512 bytes and transferred 1 bytes of data,why it is showing this error..

![Screenshot from 2024-10-07 18-02-48](https://github.com/user-attachments/assets/15f74cee-5c7c-477a-8491-55a2820d6ed9)

**Reply by JesusPoderoso** [reply]
We will take a look at it in the following days and come back with some feedback @Employeeprojec

**Comment by JesusPoderoso** [comment]
Hi @Employeeprojec, thanks for using Fast DDS.
The [shared memory transport](https://fast-dds.docs.eprosima.com/en/latest/fastdds/transport/shared_memory/shared_memory.html#transport-sharedmemory-sharedmemory) segment size close to or smaller than the data size poses a high risk of data loss, since the write operation will overwrite the buffer during a single send operation.
Ensure you configure the segment size to be big enough to allocate your data type.

**Comment by EugenioCollado** [comment]
I am closing this conversation as it has already been answered and the corresponding code improvements are described [here](https://github.com/eProsima/Fast-DDS/pull/5464) and [here](https://github.com/eProsima/Fast-DDS/pull/5473).
